### ISSUE TITLE
CTD when accepting King's proposal to help with war (Royal Frigate + troops)

### ISSUE THREAD

**Issue by nci-wtp** [We-the-People-civ4col-mod/Mod#854]
This happened in a number of my games. The issue is 100% reproducible with the save file attached.

[AutoSave_AD-1517-January.ColonizationSave.txt](https://github.com/We-the-People-civ4col-mod/Mod/files/11465824/AutoSave_AD-1517-January.ColonizationSave.txt)

**Comment by Nightinggale** [comment]
```cpp
CvPlayer::handleDiploEvent()
case DIPLOEVENT_ROYAL_INTERVENTION:
CvCity *locationToAppear = kPlayer.buyUnitFromParentPlayer(getID(), "UNITCLASS_ROYAL_INTERVENTIONS_SHIP", 1, "", 0, LocationFlags::LocationFlagDeepCoastal, false, false);
```
locationToAppear is NULL, yet it is used without checking for NULL, specifically it tries to print the name of the city.

Same savegame has an assert failure in CvPlayer::changeUnitClassMaking where the number for bishops is changed to -1.

Right now I have more questions than answers, but at least this bug can be easily reproduced and as such testing is less of an issue than it has been with some other recent bug reports.

**Comment by nci-wtp** [comment]
At least on some occasions with this bug, I was offered King's help for less than it's nominal value of 8550 gold. Maybe this information can be helpful...

**Comment by Nightinggale** [comment]
Fixed the crash. The new ship is now placed in Europe rather than crashing on a NULL city.

I found one more bug in this savegame. First the king asks you to make peace with the French and regardless of your answer, then next thing the king says is the war is dragging on and he offers to sell units. This means if you accept peace, he will offer units for the war, which isn't there. This in turn seems to make the DLL set the enemy player to player 0 (yourself), through I'm not sure that hurts anything other than possibly log entries.

In short while the originally reported bug has been fixed, the savegame is still valid for bug hunting.

**Comment by nci-wtp** [comment]
@Nightinggale in another game, I received only a ship, and no troops, when ship was placed in Europe. 

The issue is reproducible with the save file in first post.

**Comment by raystuttgart** [comment]
Are you really talking about **WTP 4.1** (develop)?

Because the "Royal Interventions" should **NOT even trigger** in case there is no "deep water coast".
There is **no such logic** that it should in that case try to place any Ship in **Europe** ...

The only thing that may try to place Ships in Europe are the Python Events.
(No DLL Diplo Event should do so.)

**Comment by raystuttgart** [comment]
**Edit:**
I see that somebody rewrote my logic in some way.
So I just do not know what happens anymore.

This code was definitely not written by me ...
But I also see no comments so I do not know who wrote it.

```
				LocationFlags location;
				location.deepCoastal = true;
				location.europe = true;
```

Generally the logic for these DLL Diplo Dialogues should simply never trigger (as there are **already safety checks** for that) in case the Player has no **Deep Harbor Coast**.

**Comment by raystuttgart** [comment]
Also I am really not sure why it spawns Royal Ships and Soliders **in Europe** then instead of simply on the **Starting Plot**.
(Which would have been the much much easier and less confusing solution.)

Ah ok, I see.
**@Nightinggale** implemented the spawning in Europe on purpose.

**Comment by nci-wtp** [comment]
@raystuttgart 

> Are you really talking about WTP 4.1 (develop)?

Of course, updated couple hours ago. 

> Generally the logic for these DLL Diplo Dialogues should simply never trigger (as there are **already safety checks** for that) in case the Player has no **Deep Harbor Coast**.

That would be unreasonably punishing for players without cities with (deep) coast tiles, don't you think?

**Comment by raystuttgart** [comment]
I was just **explaining what my old logic did** or at least was supposed to do - not trying to justify if it was wrong or right.
But right now it behaves **completely different** than what I had implemented / known ... thus I am surprised and cannot really help. 

**Comment:**
Spawning the Units on Starting Plot may still be a lot easier and safer than in Europe.
But ok, Europe is also fine for me.
